Feat/sigimm 45  rebuild jobs

sig-mmiddleton opened the pull request:
Rebuild the solr index job to be more stable.
This has mostly involved porting the functionality of the existing index task into a queued job and leveraging the steps feature of queued jobs.
Have added two variants of this task, one of which will clear existing data from the solr index before running.
I have left the existing task in place, as i'm not sure yet whether we wish to keep it or remove it in favour of the index job. For the moment at least, I think it will helpful for testing to run them both in one environment and compare.

See #7 for further info.

sig-mmiddleton committed:
feat: rebuild SolrIndexTask as a Queued Job #7

This should function essentially the same as the SolrIndexTask did,
but leveraging the steps functionality of Queued Jobs

sig-mmiddleton committed:
feat: add variant of index job that does not clear out index before running #7

sig-mmiddleton committed:
feat: refine index logging messages #7

sig-shadae commented on src/Jobs/FullSolrIndexJob.php:
```diff
@@ -0,0 +1,390 @@
+ *
+ * It always runs on all indexes, to make sure all indexes are up to date.
+ *
+ * It will clear our any existing index data before running.
```
Typo here - "our"

sig-shadae commented on src/Jobs/FullSolrIndexJob.php:
```diff
@@ -0,0 +1,390 @@
+    protected $indexes;
+
+    /**
+     * @var BaseIndex Current core being indexed
```
`@var BaseIndex` should be on a separate line

sig-shadae commented on src/Jobs/FullSolrIndexJob.php:
```diff
@@ -0,0 +1,390 @@
+    protected $shouldClearIndex = true;
+
+    /**
+     * My name
```
This doesn't exactly describe what this function does. Can you reword this?

sig-shadae commented on src/Jobs/FullSolrIndexJob.php:
```diff
@@ -0,0 +1,390 @@
+        return 'Index groups to Solr search';
+    }
+
+    public function setup()
```
Missing docblock

sig-shadae commented on src/Jobs/FullSolrIndexJob.php:
```diff
@@ -0,0 +1,390 @@
+     * Index a group of a class for a specific state and index
+     *
+     * @param string $group Group to index
+     * @param string $class Class to index
```
Missing param `$index` from docblock

sig-shadae commented on src/Jobs/FullSolrIndexJob.php:
```diff
@@ -0,0 +1,390 @@
+     *
+     * @codeCoverageIgnore This is actually tested through reflection
+     * @param string $index Index that is currently running
+     * @param int $group Group currently attempted to index
```
Param `$group` in docblock not used as function param

sig-shadae commented on src/Jobs/FullSolrIndexJob.php:
```diff
@@ -0,0 +1,390 @@
+    /**
+     * Set array of data to index
+     *
+     * @param array $classToIndex
```
Param `$classToIndex` in docblock not used as function param

sig-shadae commented on src/Jobs/FullSolrIndexJob.php:
```diff
@@ -0,0 +1,390 @@
+     * @param array $classToIndex
+     * @return FullSolrIndexJob
+     */
+    public function setIndexableData($indexableData)
```
Param `$indexableData` not in docblock

sig-shadae commented on src/Jobs/FullSolrIndexJob.php:
```diff
@@ -0,0 +1,390 @@
+    /**
+     * Set an instance of SolrCoreService
+     *
+     * @param SolrCoreService $batchLength
```
Param `$batchLength` in docblock not used as function param

sig-shadae reviewed (changes requested):
Great job with this, just some documentation updates needed.
Questions/points:
1. Are all the functions within the FullSolrIndexJob class required? Some look to be unused
2. I wonder if there should be an message logged when the index is complete

sig-mmiddleton committed:
docs: update docblocks and formatting #7

sig-mmiddleton committed:
Merge branch 'sigimm-55--deployment-plan' into feat/sigimm-45--rebuild-jobs

sig-mmiddleton committed:
feat: strip out functionality from old index task in favour of index job #7

sig-mmiddleton committed:
feat: add additional index task for triggering a full reindex #7

sig-mmiddleton committed:
docs: update documentation to reflect new index task #7

sig-mmiddleton committed:
feat: revert batch length change #7

sig-mmiddleton commented:
Have added further changes:
Removed old index task functionality - now the index task will simply queue the job.
Added a second task for queuing the job variant.
Updated documentation.

sig-shadae commented on src/Jobs/FullSolrIndexJob.php:
```diff
@@ -0,0 +1,395 @@
+<?php
+/**
+ * Class FullSolrIndexJob|Firesphere\SolrSearch\Jobs\SolrIndexTask Index Solr cores
```
Classname needs updating

sig-shadae commented on src/Jobs/FullSolrIndexJob.php:
```diff
@@ -0,0 +1,395 @@
+ * @author Simon `Firesphere` Erkelens; Marco `Sheepy` Hermo
+ * @copyright Copyright (c) 2018 - now() Firesphere & Sheepy
+ * @author Signify Ltd <info@signify.co.nz>
+ * Signify Ltd modified code in Aug 2024
```
Is the information above correct? This looks to be a new class

sig-mmiddleton replied:
I copied over the existing author/copyright information as most of the content of this file is directly pulled from the old index task. I figured it made sense to do so.

sig-shadae commented on src/Jobs/FullSolrIndexJob.php:
```diff
@@ -0,0 +1,395 @@
+    {
+        // Generate filtered list of local records
+        $baseClass = DataObject::getSchema()->baseDataClass($class);
+        /** @var DataList|DataObject[] $items */
```
Missing import

sig-shadae commented on src/Jobs/FullSolrIndexJob.php:
```diff
@@ -0,0 +1,395 @@
+    /**
+     * Execute the update on the client
+     *
+     * @param SS_List $items Items to index
```
Missing import

sig-shadae commented on src/Jobs/FullSolrIndexJob.php:
```diff
@@ -0,0 +1,395 @@
+    {
+        // Generate filtered list of local records
+        $baseClass = DataObject::getSchema()->baseDataClass($class);
+        /** @var DataList|DataObject[] $items */
```
Missing import

sig-shadae commented on src/Jobs/SolrIndexJob.php:
```diff
@@ -5,215 +5,27 @@
  *
  * It always runs on all indexes, to make sure all indexes are up to date.
  *
+ * It will not clear our any existing index data before running.
```
Typo "our"

sig-shadae commented on src/Jobs/SolrIndexJob.php:
```diff
@@ -5,215 +5,27 @@
  * @package Firesphere\Solr\Search
  */
-class SolrIndexJob extends AbstractQueuedJob
+class SolrIndexJob extends FullSolrIndexJob
```
Missing import

sig-mmiddleton replied:
This import isn't required, as it is in the same namespace

sig-shadae commented on src/Tasks/FullSolrIndexTask.php:
```diff
@@ -0,0 +1,43 @@
+     */
+    private static $segment = 'FullSolrIndexTask';
+    /**
+     * @var string $title
```
Is `$title` needed in this docblock?

sig-mmiddleton replied:
Probably excessive, but it matches the parent class.

sig-shadae commented on src/Tasks/FullSolrIndexTask.php:
```diff
@@ -0,0 +1,43 @@
+    private static $segment = 'FullSolrIndexTask';
+    /**
+     * @var string $title
+     * Shown in the overview on the {@link TaskRunner}
```
Missing import

sig-shadae commented on src/Tasks/FullSolrIndexTask.php:
```diff
@@ -0,0 +1,43 @@
+     */
+    protected $title = 'Full Solr Index update';
+    /**
+     * @var string $description Describe the implications the task has,
```
Is `$description` needed in this docblock?

sig-mmiddleton replied:
Probably excessive, but it matches the parent class.

sig-shadae commented on src/Tasks/SolrIndexTask.php:
```diff
@@ -43,410 +25,37 @@
-     *
-     * @var string
+     * @var string $title
+     * Shown in the overview on the {@link TaskRunner}
```
Missing import

sig-shadae reviewed (changes requested):
Couple more changes,
Also, looks to be some inconsistencies with function docblocks and strong typing, specifically when the functions return `void` where the function will only use one or the other.

sig-mmiddleton committed:
docs: update documentation and imports #7

sig-shadae reviewed (changes requested):
I've had a look through the documentation, I wonder if a comment is needed in 01-Installation.md that the script at line 36 instead queues the job to add documents to your index, as the task does not directly do this with these changes.
Additionally, both FullSolrIndexJob and SolrIndexJob can be queued at the same time. When looking at the CMS queued jobs, both are titled the same so not easily distinguishable.

sig-mmiddleton committed:
feat: distinguish title of different index jobs #7

sig-mmiddleton committed:
docs: update documentation regarding index task #7

sig-mmiddleton commented:
Have differentiated the display title of the index jobs and updated the documentation relating to the tasks.
Review re-requested.

sig-shadae reviewed (approved):
Looks good to me :)
